pkg/alertmanager/distributor/distributor.go

```diff
@@ -0,0 +1,283 @@
+	d.subservicesWatcher.WatchManager(d.subservices)
+
+	d.Service = services.NewBasicService(d.starting, d.running, d.stopping)
+	return d, nil
+}
+
+func (d *Distributor) initMetrics(r prometheus.Registerer) {
+	d.receivedRequests = promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_alertmanager_distributor_received_requests_total",
+		Help: "The total number of requests received.",
+	}, []string{"user"})
+	d.amSends = promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_alertmanager_distributor_alertmanager_send_total",
+		Help: "The total number of requests sent to alertmanager.",
+	}, []string{"ingester"})
+	d.amSendFailures = promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_alertmanager_distributor_alertmanager_send_failures_total",
+		Help: "The total number of requests failed to send to alertmanager.",
+	}, []string{"ingester"})
+	d.replicationFactor = promauto.With(r).NewGauge(prometheus.GaugeOpts{
+		Name: "cortex_alertmanager_distributor_replication_factor",
+		Help: "The configured replication factor.",
+	})
+}
+
+func (d *Distributor) starting(ctx context.Context) error {
+	return services.StartManagerAndAwaitHealthy(ctx, d.subservices)
+}
+
+func (d *Distributor) running(ctx context.Context) error {
+	<-ctx.Done()
+	d.requestsInFlight.Wait()
+	return nil
+}
+
+func (d *Distributor) stopping(_ error) error {
+	return services.StopManagerAndAwaitStopped(context.Background(), d.subservices)
+}
+
+// IsPathSupported returns true if the given route is currently supported by the Distributor.
+// TODO: This will go away in future after we gradually add support for the entire API.
+func (d *Distributor) IsPathSupported(path string) bool {
+	// API can be found at https://petstore.swagger.io/?url=https://raw.githubusercontent.com/prometheus/alertmanager/master/api/v2/openapi.yaml.
+	return strings.HasSuffix(path, "/alerts") ||
+		strings.HasSuffix(path, "/alerts/groups")
+}
+
+// DistributeRequest shards the writes and returns as soon as the quorum is satisfied.
+// In case of reads, it proxies the request to one of the alertmanagers.
+// DistributeRequest assumes that the caller has verified IsPathSupported return
+// true for the route.
+func (d *Distributor) DistributeRequest(w http.ResponseWriter, r *http.Request) {
+	//TODO: Initialise tracing here.
+
+	d.requestsInFlight.Add(1)
+	defer d.requestsInFlight.Done()
+
+	userID, err := tenant.TenantID(r.Context())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	d.receivedRequests.WithLabelValues(userID).Inc()
+
+	if r.Method == http.MethodGet {
+		d.doRead(userID, w, r)
+	} else {
+		d.doWrite(userID, w, r)
+	}
+}
+
+func (d *Distributor) doWrite(userID string, w http.ResponseWriter, r *http.Request) {
+	source := util.GetSourceIPsFromOutgoingCtx(r.Context())
+
+	var body []byte
+	var err error
+	if r.Body != nil {
+		body, err = ioutil.ReadAll(r.Body)
```

Comment:
We already read the body as we do `server.HTTPRequest(r)`, so any errors we might expect from it can be handled on that error condition. Is there a reason why we need to do it here?

Reply:
When we have multiple requests spawning form here, only one of them will be able to read it full and will cause panics in others. So we copy the body here to be able to use in multiple requests.